burger/toppings/packetinstructions.py

```diff
@@ -794,19 +812,44 @@ def _handle_3_arg_buffer_call(classloader, classes, instruction, verbose,
                                         type="Map.Entry<" + key_type + ", " + val_type + ">",
                                         var="itv", value="it.next()"))
             operations += _PIT._lambda_operations(
-                classloader, classes, instruction, verbose,
+                classloader, classes, instruction.pos, verbose,
                 key_info, [instance, "itv.getKey()"]
             )
             # TODO: Does the SUB_INS_EPSILON work correctly here?
             # I think this will lead to [1.01, 1.02, 1.03, 1.01, 1.02, 1.03]
             # which would get sorted wrongly, but I'm not sure
             operations += _PIT._lambda_operations(
-                classloader, classes, instruction, verbose,
+                classloader, classes, instruction.pos, verbose,
                 val_info, [instance, "itv.getValue()"]
             )
             # Same jank as with the one in _handle_2_arg_buffer_call
             operations.append(Operation(instruction.pos + 1 - SUB_INS_EPSILON, "endloop"))
             return operations
+        elif desc.args[0].name == "com/mojang/datafixers/util/Either":
+            # Write a boolean indicating whether it's left, and then call the correct consumer
+            # with the packetbuffer and value.
+            operations = []
+            field = args[0]
+            left_consumer = args[1]
+            right_consumer = args[2]
+            assert isinstance(field, StackOperand)
+            assert isinstance(left_consumer, InvokeDynamicInfo)
+            assert isinstance(right_consumer, InvokeDynamicInfo)
+            operations.append(Operation(instruction.pos, "write", type="boolean",
+                                        field=field.value + ".isLeft()"))
+            operations.append(Operation(instruction.pos, "if",
+                                        condition=field.value + ".isLeft()"))
+            operations += _PIT._lambda_operations(
+                classloader, classes, instruction.pos, verbose,
+                left_consumer, [instance, field.value + ".left()"]
+            )
+            operations.append(Operation(instruction.pos + 1 - SUB_INS_EPSILON, "else"))
+            operations += _PIT._lambda_operations(
+                classloader, classes, instruction.pos + 1, verbose,
+                right_consumer, [instance, field.value + ".right()"]
+            )
+            operations.append(Operation(instruction.pos + 2 - SUB_INS_EPSILON, "endif"))
```

Comment:
When I didn't add +1 to the instruction pos the `else` was empty and the instructions were getting mixed with the `if` instructions, so I think the implementation for Map might actually be wrong. I also just checked by doing `print(instruction.pos)` at the beginning of a loop where a function is disassembled, and it does look like they're byte offsets since the positions usually increase by more than 1 at a time.

Ok also you don't need to answer this, but why do operations even need to store their position? Can't that just be figured out from the order in the operations list?

Reply:
In that case, it's probably fine.

> Ok also you don't need to answer this, but why do operations even need to store their position? Can't that just be figured out from the order in the operations list?

I'm pretty sure that specific part of the code predates me. I do know that they aren't initially sorted (which is what `ordered_operations` does). I assume that it's related to `endif`, which I think gets put into the list as soon as a jump is seen. Not that the if handling is great; see Pokechu22#2 and Pokechu22#3. I think it also relates to converting from a list of instructions into a nested tree structure (see `format`).